src/main/java/net/neoforged/neoforge/transfer/resources/ItemResource.java

```diff
@@ -0,0 +1,280 @@
+    /**
+     * Codec for an item resource. Same format as {@link #CODEC}, and also accepts empty resources.
+     */
+    public static final Codec<ItemResource> OPTIONAL_CODEC = Codec.lazyInitialized(() -> ExtraCodecs.optionalEmptyMap(CODEC).xmap(ItemResource::fromOptional, ItemResource::asOptional));
+
+    @SuppressWarnings("OptionalUsedAsFieldOrParameterType")
+    public static ItemResource fromOptional(Optional<ItemResource> optional) {
+        return optional.orElse(ItemResource.EMPTY);
+    }
+
+    public Optional<ItemResource> asOptional() {
+        return isEmpty() ? Optional.empty() : Optional.of(this);
+    }
+
+    /**
+     * Stream codec for an item resource. Accepts empty resources.
+     */
+    public static final StreamCodec<RegistryFriendlyByteBuf, ItemResource> STREAM_CODEC = StreamCodec.composite(
+            ByteBufCodecs.holderRegistry(Registries.ITEM), ItemResource::getHolder,
+            DataComponentPatch.STREAM_CODEC, ItemResource::getComponentsPatch,
+            ItemResource::of);
+
+    public static ItemStack itemStackOf(IResourceStack<ItemResource> resourceStack) {
+        return resourceStack.resource().toStack(resourceStack.amount());
+    }
+
+    public static final ItemResource EMPTY = new ItemResource(ItemStack.EMPTY);
+    public static final ResourceStack<ItemResource> EMPTY_STACK = ResourceStack.of(EMPTY, 0);
+    public static final MutableResourceStack<ItemResource> EMPTY_MUTABLE_STACK = MutableResourceStack.of(EMPTY, 0);
+
+    /**
+     * This is used only for registry, you should not use this method!
+     */
+    @ApiStatus.Internal
+    public static ItemResource invalidateDefault(ItemLike item) {
+        if (item.asItem() == Items.AIR) return EMPTY;
+        return new ItemResource(item.asItem().getDefaultInstance().copyWithCount(1));
+    }
+
+    public static ItemResource of(ItemStack itemStack) {
+        if (itemStack.isEmpty()) return EMPTY;
+
+        if (itemStack.isComponentsPatchEmpty())
+            return itemStack.getItem().getDefaultResource();
+
+        return new ItemResource(itemStack.copyWithCount(1));
+    }
+
+    /**
+     * <strong>Note:</strong> This cannot be called before your item is registered
+     */
+    public static ItemResource of(ItemLike item) {
+        if (item.asItem() == Items.AIR) return EMPTY;
+        return item.asItem().getDefaultResource();
+    }
+
+    /**
+     * <strong>Note:</strong> This cannot be called before your item is registered
+     */
+    public static ItemResource of(Holder<Item> item, DataComponentPatch patch) {
+        if (item.value() == Items.AIR) return EMPTY;
+        if (patch.isEmpty()) return item.value().getDefaultResource();
+        return item.value().getDefaultResource().withPatch(patch);
+    }
+
+    /**
+     * <strong>Note:</strong> This cannot be called before your item is registered
+     */
+    public static ItemResource of(Holder<Item> item) {
```

Comment:
Given this doesn't make it lazy and you aren't bothering to check if it is empty by resource key, just have the impl of this method be `return of(item.value());`